test.py

```diff
@@ -0,0 +1,595 @@
+                logger.info("For now, proceeding with prayer")
+                # TODO
+                return True
+        if 'CCM_AUTH_TOKEN' in os.environ:
+            logger.info("ccm auth token provided.")
+            return True
+        else:
+            logger.info("No CLUSTER_URL or CCM_AUTH_TOKEN provided. FAIL")
+            return False
+
+    def have_gopath():
+        if 'GOPATH' in os.environ:
+            logger.info("GOPATH is ... set.")
+            return True
+        else:
+            logger.info("GOPATH is ... not set.  FAIL")
+            return False
+
+    def have_ssh_key():
+        try:
+            completed_cmd = subprocess.run(['ssh-add', '-l'], stdout=subprocess.PIPE)
+            if completed_cmd.returncode != 0:
+                logger.info("ssh-add invocation returned failure.  FAIL")
+                return False
+            if b'SHA256:i+SOiM8V5+yI3C0LoNlPaDk+ffCdOk2ZmDWoRIK8aw4' in completed_cmd.stdout:
+                logger.info("ssh-agent .. has ec2 key.")
+                return True
+            else:
+                logger.info("ssh-agent .. does not have ec2 key.  FAIL")
+                return False
+        except:
+            logger.info("ssh-add invocation failed.  FAIL")
+            return False
+
+    results = {}
+
+    # build requirements
+    results['java'] = have_command("java")
+    results['go'] = have_command("go")
+    results['gopath'] = have_gopath()
+    # TODO: verify libssl-dev or equivalent installed
+    # In other words: #include <openssl/opensslv.h> should work
+
+    if sys.platform != 'darwin':
+        results['upx'] = have_command("upx")
+
+    # upload requirements
+    results['aws'] = have_command("aws")
+    # TODO: verify can access our s3 bucket
+
+    results['docker'] = have_command("docker")
+    if results['docker']:
+        results['docker_works'] = docker_works()
+    # TODO: validate we have the docker access
+
+    # test requirements
+    results['virtualenv'] = have_command("virtualenv")
+    results['cluster'] = have_or_can_create_cluster()
+    results['ec2_ssh_key'] = have_ssh_key()
+    results['jq'] = have_command("jq")
+
+    failures = [v for v in results.values() if v==False]
+    if failures:
+        msg = "Requirements not met."
+        logger.info(msg)
+        raise TestRequirementsNotMet(msg, results)
+    logger.info("All (tested) requirements met.")
+    return results
+
+def get_cluster():
+    "Bring up a cluster and return, or return an already running cluster"
+    pass
+
+def setup_frameworks(run_attrs):
+    if not run_attrs.test:
+        fwinfo.autodiscover_frameworks()
+    else:
+        for framework in run_attrs.test:
+            fwinfo.add_framework(framework)
+
+    if run_attrs.order == "random":
+        fwinfo.shuffle_order()
+
+    fw_names = fwinfo.get_framework_names()
+    logger.info("Frameworks initialized: %s", ", ".join(fw_names))
+
+
+def _action_wrapper(action_name, framework, function, *args):
```

Comment:
Maybe you could explain this one in person.  The scope is very small and the chain is one deep, so I'm not sure what you mean.

This is
1 remember we started doing this thing
2 try to do the thing
3 remember whether the thing worked or failed

Yeah, this could be inlined, or we could have 4 nearly identical functions, but neither seems better to me.